subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/SelectiveCompiler.java

```diff
@@ -55,6 +55,11 @@ public SelectiveCompiler(PreviousCompilation previousCompilation,
 
     @Override
     public WorkResult execute(T spec) {
+        if (spec.getSourceRoots().isEmpty()) {
+            LOG.info("Full recompilation is required because the source roots could not be inferred.");
```

Comment:
Isn't this even worse? Doesn't that mean that we should disable incremental compilation altogether?

Reply:
I may misunderstand you, but this is just moving [this code](https://github.com/gradle/gradle/blob/7e7b0e729a75236bc3305c5fb5c6bd2aa53b3403/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalCompilerDecorator.java#L74).

Reply:
Yes, I understand. Though I don't think the problem of not being able to infer source roots will go away easily. Let's leave this in place for Java compilation. For Groovy compilation we can't end up here, since we disable incremental compilation if we can't resolve the source roots, right?

Reply:
Exactly.